book/src/guide/specifications.md

```diff
@@ -532,9 +694,71 @@ TODO
 
 ## Extern Specs
 
+Sometimes you may want to refine a struct or function that outside your code. We
+refer to such a specification as an "extern spec," which is short for "external
+specification."
+
+Currently, Flux supports extern specs for functions, structs, enums, traits and impls.
+The support is a bit rudimentary. For example, multiple impls for a struct (such as `&[T]`
+and `[T]`) may conflict, and extern specs for structs only support opaque refinements.
+
+Extern specs are given using `extern_spec` attribute macro, which is provided
+by the procedural macros package `flux_rs`.
+
+```
+use flux_rs::extern_spec;
+```
+
+
 The `extern_spec` is used to provide `flux` signatures for functions defined in _external_ crates. See the [specifications guide](specs.md) for more details.
 
-### Functions
+### Extern Functions
+
+An example of refining an extern function can be found
+[here](https://github.com/flux-rs/flux/blob/d49a74dc59b2b9bb1dda01ee019d0ab9a66cdd89/flux-tests/tests/pos/surface/extern_spec_macro.rs).
+
+To define an extern spec on a function, you need to do three things, which
+happen to correspond to each of the below lines.
+
+```rust
+{{#include ../../../tests/tests/neg/surface/extern_spec_swap.rs:5:7}}
+```
+
+1. Add the `#[extern_spec]` attribute. This attribute optionally takes a path;
+   in the above example, this is `std::mem`. You can use this path to qualify
+   the function. So in the above example, the function we are targeting has the
+   full path of `std::mem::swap`.
+2. Add a `#[spec(...)]` (or equivalently `#[flux_rs::sig(...)]`) attribute, which
+   is required for any extern spec on a function. This signature behaves as if
+   the `#[flux_rs::trusted]` attribute was added, because we cannot _actually_
+   check the implementation. Instead, flux just verifies some simple things,
+   like that the function arguments have compatible types.
+3. Write a function stub whose rust signature matches the external function.
+
+If you do the above, you can use `std::mem::swap` as if it were refined by the
+above type.
+
+
+```rust
+{{#include ../../../tests/tests/neg/surface/extern_spec_swap.rs:9:16}}
+```
+
+You shouldn't need to know the details, but if you're curious,
+here's how the macro works. In the code above, flux parses the
+`std::mem` into a module path and then transforms the function into
+
+```rust
+#[extern_spec]
+#[spec(fn(x: &mut T[@vx], y: &mut T[@vy]) ensures x: T[vy], y: T[vx])]
+#[allow(unused, dead_code)]
+fn __flux_extern_spec_swap<T>(x: &mut T, y: &mut T) {
+    std::mem::swap(x, y)
+}
+```
```

Comment:
We can probably get rid of this.

```suggestion
You shouldn't need to know the details, but if you're curious,
here's how the macro works. In the code above, flux parses the
`std::mem` into a module path and then transforms the function into

```rust
#[extern_spec]
#[spec(fn(x: &mut T[@vx], y: &mut T[@vy]) ensures x: T[vy], y: T[vx])]
#[allow(unused, dead_code)]
fn __flux_extern_spec_swap<T>(x: &mut T, y: &mut T) {
    std::mem::swap(x, y)
}
```
```

Reply:
So to clarify: delete the whole discussion? The "suggestion" on github is  no-op (as in the suggestion seems to be the original text itself...?)

Reply:
ups, yeah, I think we can remove this, no need to "stabilize" the expansion.